src/main/scala/miss/supervisor/Supervisor.scala

```diff
@@ -27,66 +31,101 @@ class Supervisor(config: Config) extends FSM[State, Data] {
+        goto(WarmUp) using SupervisorData(newWorkers, areaActors, None, 0, 0)
       }
       else {
         goto(WaitingForWorkers) using WorkersData(newWorkers)
       }
   }
 
+  when(WarmUp) {
+    case Event(TimeFrameUpdate(x, y, newTimeFrame), SupervisorData(workers, areaActors, cityVisualizer, firstComputedFrame, lastComputedFrame)) =>
+      cityVisualizer.foreach(ar => ar ! CityVisualizationUpdate(x, y, newTimeFrame))
+      val currentTimeFrame = if (newTimeFrame > lastComputedFrame) newTimeFrame else lastComputedFrame
+      goto(WarmUp) using SupervisorData(workers, areaActors, cityVisualizer, firstComputedFrame, currentTimeFrame)
+    case Event(WarmUpDone, SupervisorData(workers, areaActors, cityVisualizer, firstComputedFrame, lastComputedFrame)) =>
+      log.info("Starting simulation phase")
+      setTimer("simulation-done-timer", SimulationDone, simulationTimeSeconds seconds)
+      goto(Working) using SupervisorData(workers, areaActors, cityVisualizer, lastComputedFrame, lastComputedFrame)
+  }
+
   when(Working) {
-    case Event(StartVisualization(x, y), SupervisorData(_, actors, _)) =>
+    case Event(StartVisualization(x, y), SupervisorData(_, actors, _, _, _)) =>
       val visualizer = sender()
       actors(x)(y) ! VisualizationStartRequest(visualizer)
       stay
-    case Event(StopVisualization(x, y), SupervisorData(_, actors, _)) =>
+    case Event(StopVisualization(x, y), SupervisorData(_, actors, _, _, _)) =>
       val visualizer = sender()
       actors(x)(y) ! VisualizationStopRequest(visualizer)
       stay
-    case Event(CityVisualizationStartRequest, SupervisorData(workers, areaActors, _)) =>
-      goto(Working) using SupervisorData(workers, areaActors, Some(sender()))
-    case Event(CityVisualizationStopRequest, SupervisorData(workers, areaActors, _)) =>
-      goto(Working) using SupervisorData(workers, areaActors, None)
-    case Event(TimeFrameUpdate(x, y, newTimeFrame), SupervisorData(_, _, cityVisualizer)) =>
+    case Event(CityVisualizationStartRequest, SupervisorData(workers, areaActors, _, firstComputedFrame, lastComputedFrame)) =>
+      goto(Working) using SupervisorData(workers, areaActors, Some(sender()), firstComputedFrame, lastComputedFrame)
+    case Event(CityVisualizationStopRequest, SupervisorData(workers, areaActors, _, firstComputedFrame, lastComputedFrame)) =>
+      goto(Working) using SupervisorData(workers, areaActors, None, firstComputedFrame, lastComputedFrame)
+    case Event(TimeFrameUpdate(x, y, newTimeFrame), SupervisorData(workers, areaActors, cityVisualizer, firstComputedFrame, lastComputedFrame)) =>
       cityVisualizer.foreach(ar => ar ! CityVisualizationUpdate(x, y, newTimeFrame))
-      stay
+      val currentTimeFrame = if (newTimeFrame > lastComputedFrame) newTimeFrame else lastComputedFrame
+      goto(Working) using SupervisorData(workers, areaActors, cityVisualizer, firstComputedFrame, currentTimeFrame)
```

Comment:
Actually it can be `stay using ...` when you do not change state, but change data only.

Please check for it in other places as well.